src/vmm/src/irq_allocator.rs

```diff
@@ -0,0 +1,84 @@
+use std::fmt;
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// An irq allocator which gives next available irq.
+/// It is mainly used for non-legacy devices.
+// There are a few reserved irq's on x86_64. We just skip all the inital
+// reserved irq to make the implementaion simple. This could be later extended
+// to cater more complex scenario.
+#[derive(Debug)]
+pub struct IrqAllocator {
+    // Tracks the last allocated irq
+    last_used_irq: u32,
+    max: u32,
+}
+
+impl IrqAllocator {
+    pub fn new(last_used_irq: u32, max: u32) -> Result<Self> {
```

Comment:
For the interface it doesn't make sense for this parameter to be called `last_used_irq` because the caller typically wants to initialize a range of available IRQs. Names such as `first_irq` and `last_irq` are more appropriate for this initialization. It is also more clear that with these we mean the actual value, and not a size. From that point of view `max` is an ambiguous name because it's unclear whether it's the irq, or the size of the interval.

You can keep the name as `last_used_irq` in the field name because that's the intention with which we're going to use it.

Reply:
I changed max to `last_irq` but for `last_used_irq` I think because the user can never get lets say 4 so it's like what he actually last allocated.